### PR TITLE
feat: Robustness against prover crash on (reset)

### PR DESCRIPTION
**Pull request by MikaelMayer**
This PR provides backward compatible support for Z3 4.8.5 on Windows, which sometimes would crash when receiving the `(reset)` command - this happened in the Dafny test suite (https://github.com/dafny-lang/dafny/pull/1540).
This bug was corrected in Z3 4.8.8, but I believe this feature can be our defense-in-depth strategy for dealing with any unexpected Z3 reset crash in the future.

## Changes

- `SMTLibProcessTheoremProver.cs` Added a check right after the `(reset)` command via a `HasProverDied()` command to the prover. If it was not the first (reset) sent, it restarts a new fresh process.
- `SMTLibProcess.cs` Added the method `HasProverDied()` which invokes `PingPong` and returns the exception if one was thrown.

### CODE REVIEW

File: Source/Provers/SMTLib/SMTLibProcessTheoremProver.cs [hunk 1]
```diff
@@ -129,9 +130,9 @@ private void SetupAxiomBuilder(VCExpressionGenerator gen)
       }
     }
 
-    void SetupProcess()
+    void SetupProcess(bool recover = false)
```
Comment (keyboardDrummer) [inline Source/Provers/SMTLib/SMTLibProcessTheoremProver.cs:133]: It seems like you don't need the additional `recover` argument if you run `Process = null;` before calling `SetupProcess` on line 660.

File: Source/Provers/SMTLib/SMTLibProcessTheoremProver.cs [hunk 2]
```diff
@@ -644,6 +645,23 @@ public override void Reset(VCExpressionGenerator gen)
...
     }
 
+    private void RecoverIfProverCrashedAfterReset()
+    {
+      resetCount += 1;
+      if (Process.HasProverDied() is { } e)
+      {
+        if (resetCount == 1)
+        {
+          // First-time reset cannot be recovered
```
Comment (keyboardDrummer) [inline Source/Provers/SMTLib/SMTLibProcessTheoremProver.cs:655]: Why not?

File: Source/Provers/SMTLib/SMTLibProcessTheoremProver.cs [hunk 3]
```diff
@@ -644,6 +645,23 @@ public override void Reset(VCExpressionGenerator gen)
       }
     }
 
+    private void RecoverIfProverCrashedAfterReset()
+    {
+      resetCount += 1;
+      if (Process.HasProverDied() is { } e)
```
Comment (keyboardDrummer) [inline Source/Provers/SMTLib/SMTLibProcessTheoremProver.cs:651]: I think I would find `is Exception e` easier to read than `is { } e`

File: Source/Provers/SMTLib/SMTLibProcess.cs [hunk 4]
```diff
@@ -125,6 +125,20 @@ public bool IsPong(SExpr sx)
       return sx != null && sx.Name == ":name";
     }
 
+    public ProverDiedException HasProverDied()
```
Comment (keyboardDrummer) [inline Source/Provers/SMTLib/SMTLibProcess.cs:128]: The name of this method implies that it returns a boolean.

File: Source/Provers/SMTLib/SMTLibProcessTheoremProver.cs (hunk 2 above)
Comment (MikaelMayer) [reply Source/Provers/SMTLib/SMTLibProcessTheoremProver.cs:655]: Because at this point, only `(reset)` was sent to the solver. if it crashes there, recovering it would mean spawning a new solver instance and sending it more `(reset)` command, which could loop indefinitely.
In practice this then-branch will never be entered, but it's there to prove termination (in the unlikely case a solver is corrupted for example)

File: Source/Provers/SMTLib/SMTLibProcessTheoremProver.cs (hunk 1 above)
Comment (MikaelMayer) [reply Source/Provers/SMTLib/SMTLibProcessTheoremProver.cs:133]: I followed here the principle Don't Repeat Yourself. Looking at this pattern used twice, It thought it would be better to refactor it:
`Process = null; SetupProcess();` ==> `SetupProcess(/*recover*/true)`
I see several reasons that this enhances the readability and maintainability
- Having a statement controlling the inner if of another hidden statement is error-prone and hard to follow.
- If Process is implemented in a different way down the road, we will have to change all statements before call sites
- Process is assigned twice in the first statement (once null, once the new process), but what we want here is just discard the previous process and only one assignment can be made.

Because of the purpose of this PR is not maintainability, if you tell me that you prefer the first version, I'd happily make the change if this gets this PR accepted more easily.

File: Source/Provers/SMTLib/SMTLibProcessTheoremProver.cs (hunk 3 above)
Comment (MikaelMayer) [reply Source/Provers/SMTLib/SMTLibProcessTheoremProver.cs:651]: That's was the IDE semi-auto correction (it underlined in yellow `ProverDiedException e` when I wrote it), which was annoying.
Thanks to your suggestion, I'm writing `Exception e` and it does not complain anymore and indeed it's easier to read.

File: Source/Provers/SMTLib/SMTLibProcess.cs (hunk 4 above)
Comment (MikaelMayer) [reply Source/Provers/SMTLib/SMTLibProcess.cs:128]: Method name changed to `GetExceptionIfProverDied`

File: Source/Provers/SMTLib/SMTLibProcessTheoremProver.cs [hunk 5]
```diff
@@ -644,6 +645,23 @@ public override void Reset(VCExpressionGenerator gen)
...
+      {
+        if (resetCount == 1)
+        {
+          // First-time reset cannot be recovered
+          throw e;
+        }
+
+        Process.Close();
+        SetupProcess(true);
+        resetCount = 1;
```
Comment (keyboardDrummer) [inline Source/Provers/SMTLib/SMTLibProcessTheoremProver.cs:661]: At the start of `RecoverIfProverCrashedAfterReset` this variable is incremented, so assigning 1 to it, would mean that next time it's already 2, right? If you want to reset this value shouldn't you set it to 0 ?

File: Source/Provers/SMTLib/SMTLibProcessTheoremProver.cs (hunk 2 above)
Comment (keyboardDrummer) [reply Source/Provers/SMTLib/SMTLibProcessTheoremProver.cs:655]: Why would another reset command be sent after the recovery?

File: Source/Provers/SMTLib/SMTLibProcessTheoremProver.cs (hunk 1 above)
Comment (keyboardDrummer) [reply Source/Provers/SMTLib/SMTLibProcessTheoremProver.cs:133]: I suggest inlining the

```
if (Process != null && !recover)
      {
        return;
      }
```

Part of the method, since only one call passes `recover == false`.

File: Source/Provers/SMTLib/SMTLibProcessTheoremProver.cs (hunk 5 above)
Comment (MikaelMayer) [reply Source/Provers/SMTLib/SMTLibProcessTheoremProver.cs:661]: Next time, it's already 2, right, but there is a good reason for that.
The code is catching errors that happen right after a `(reset)` command is issued, and no other errors. Currently, Boogie sends this command when the process starts for the first time, and that's the non-recoverable `(reset)` command that would be caught and thrown again.
Other than that, if a `(reset)` throw an exception after this case, it means that the prover already did some processing and it's really after that that the `(reset)` command failed. So we just want to restart the prover and not even send a `(reset)` command that would need to be checked again.

Sorry for the confusion, I could have just put a boolean there or just remove this counter alltogether (after all, if a reset at the beginning fails, there isn't anything that we could do)

File: Source/Provers/SMTLib/SMTLibProcessTheoremProver.cs (hunk 2 above)
Comment (MikaelMayer) [reply Source/Provers/SMTLib/SMTLibProcessTheoremProver.cs:655]: You're right, actually, no `(reset)` command is sent after the recover. I think I can simplify my code then.

File: Source/Provers/SMTLib/SMTLibProcessTheoremProver.cs (hunk 1 above)
Comment (MikaelMayer) [reply Source/Provers/SMTLib/SMTLibProcessTheoremProver.cs]: That's a good point. Since there is never a point where we call `SetupProcess` "just in case", I removed all ifs and just assume that if the process was there, it should be restarted anyway with this call.

**Review by keyboardDrummer — APPROVED** [review]
